flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/recommendation/ALS.scala

```diff
@@ -581,6 +637,16 @@ object ALS {
         val userXy = new ArrayBuffer[Array[Double]]()
         val numRatings = new ArrayBuffer[Int]()
 
+        var precomputedXtX: Array[Double] = null
+
+        override def open(config: Configuration): Unit = {
+          // retrieve broadcasted precomputed XtX if using implicit feedback
+          if (implicitPrefs) {
+            precomputedXtX = getRuntimeContext.getBroadcastVariable[Array[Double]]("XtX")
+              .iterator().next()
+          }
+        }
+
         override def coGroup(left: lang.Iterable[(Int, Int, Array[Array[Double]])],
```

Comment:
I tried using a Scala `Iterable` and it won't compile. Probably because `CoGroupFunction` is a Java class that requires a Java `Iterable` in `coGroup`. One solution for avoiding Java `Iterable` would be to use a simple Scala function instead of a `CoGroupFunction` anonymous class. While that would have been feasible in the original ALS, we use a `RichCoGroupFunction` to initialize the `X^T * X` matrix, hence we need to stick with the Java `Iterable`.